### PR TITLE
UCP/AMO: Add active-message based AMO protocols

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7580/ for details (Mellanox internal link).

File: src/ucp/core/ucp_rkey.c [hunk 1]
```diff
@@ -400,23 +400,31 @@ void ucp_rkey_resolve_inner(ucp_rkey_h rkey, ucp_ep_h ep)
...
+        rkey->cache.amo_proto     = &ucp_amo_sw_proto;
+        rkey->cache.amo_rkey      = UCT_INVALID_RKEY;
+    } else {
+        rkey->cache.amo_proto     = &ucp_amo_basic_proto;
         rkey->cache.amo_rkey      = uct_rkey;
         rkey->cache.amo_proto     = &ucp_amo_basic_proto;
     }
 
-    /* If we use sw rma need to resolve destination endpoint in order to
+    /* If we use sw rma/amo need to resolve destination endpoint in order to
```
Comment (shamisp) [inline src/ucp/core/ucp_rkey.c]: at this point all lanes have to switch to SW amo

Comment (yosefe) [reply src/ucp/core/ucp_rkey.c]: if you mean the sw/hw atomicity issue, then it's enough to calculate the lane per rkey. two different atomic variables (e.g 1st in shmem heap, 2nd in .bss/.data), with diff rkeys, can use different lanes.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7584/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7588/ for details (Mellanox internal link).

File: src/ucp/core/ucp_rkey.c [hunk 2]
```diff
@@ -400,23 +400,31 @@ void ucp_rkey_resolve_inner(ucp_rkey_h rkey, ucp_ep_h ep)
                                                     UCT_MD_MEM_TYPE_HOST,
                                                     config->key.amo_lanes, rkey,
                                                     0, &uct_rkey);
-    if (rkey->cache.amo_lane != UCP_NULL_LANE) {
+    amo_sw = (rkey->cache.amo_lane == UCP_NULL_LANE);
+    if (amo_sw) {
+        rkey->cache.amo_proto     = &ucp_amo_sw_proto;
+        rkey->cache.amo_rkey      = UCT_INVALID_RKEY;
+    } else {
+        rkey->cache.amo_proto     = &ucp_amo_basic_proto;
```
Comment (brminich) [inline src/ucp/core/ucp_rkey.c]: the same line as 410 line below

File: src/ucp/rma/amo_sw.c [hunk 3]
```diff
@@ -0,0 +1,292 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2001-2017.  ALL RIGHTS RESERVED.
```
Comment (brminich) [inline src/ucp/rma/amo_sw.c]: minor: 2018

File: src/ucp/rma/amo_sw.c [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+DEFINE_AMO_SW_FOP(32)
+DEFINE_AMO_SW_FOP(64)
+
+UCS_PROFILE_FUNC(ucs_status_t, ucp_atomic_req_handler, (arg, data, length, am_flags),
+                 void *arg, void *data, size_t length, unsigned am_flags)
+{
+    ucp_atomic_req_hdr_t *atomicreqh = data;
+    ucp_worker_h worker              = arg;
+    ucp_ep_h ep                      = ucp_worker_get_ep_by_ptr(worker,
+                                                                atomicreqh->req.ep_ptr);;
```
Comment (brminich) [inline src/ucp/rma/amo_sw.c]: extra `;` in the end

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7593/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7602/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7604/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7648/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@brminich pls take a look

**Pull request merged**